src/resource_manager/gateway.py

```diff
@@ -0,0 +1,238 @@
+                "Configured gateway class %s not present on the cluster.", configured_gateway_class
+            )
+            raise CreateGatewayError(f"Gateway class {configured_gateway_class} not found.")
+
+        resource.spec["gatewayClassName"] = configured_gateway_class
+
+    @_map_k8s_auth_exception
+    def _gen_resource_from_definition(self, definition: GatewayResourceDefinition) -> dict:
+        """Generate a Gateway resource from a gateway resource definition.
+
+        Args:
+            definition: The gateway resoucre definition to use.
+
+        Returns:
+            A dictionary representing the gateway custom resource.
+        """
+        gateway = self._gateway_generic_resource(
+            apiVersion="gateway.networking.k8s.io/v1",
+            kind="Gateway",
+            metadata=ObjectMeta(
+                name=definition.gateway_name, namespace=self._namespace, labels=self._labels
+            ),
+            spec={
+                "listeners": [
+                    {
+                        "protocol": "HTTP",
+                        "port": 80,
+                        "name": f"{definition.gateway_name}-http-listener",
+                        "hostname": definition.config.external_hostname,
+                        "allowedRoutes": {"namespaces": {"from": "All"}},
+                    },
+                ]
+            },
+        )
+
+        self._set_gateway_class(
+            configured_gateway_class=definition.config.gateway_class, resource=gateway
+        )
+        logger.info("Generated gateway resource: %s", gateway)
+        return gateway
+
+    @_map_k8s_auth_exception
+    def _create_resource(self, resource: GenericNamespacedResource) -> None:
+        """Create a new gateway resource in a given namespace.
+
+        Args:
+            resource: The gateway resource object to create.
+        """
+        self._client.create(resource)
+
+    @_map_k8s_auth_exception
+    def _patch_resource(self, name: str, resource: GenericNamespacedResource) -> None:
+        """Replace an existing gateway resource in a given namespace.
+
+        Args:
+            name: The name of the resource to patch.
+            resource: The modified gateway resource object.
+        """
+        # Patch the resource with server-side apply
+        # force=True is required here so that the charm keeps control of the resource
+        self._client.patch(  # type: ignore[type-var]
+            # mypy can't detect that this is ok for patching custom resources
+            self._gateway_generic_resource,
+            name,
+            resource,
+            patch_type=PatchType.APPLY,
+            force=True,
+        )
+
+    @_map_k8s_auth_exception
+    def _list_resource(self) -> List[GenericNamespacedResource]:
+        """List gateway resources in a given namespace based on a label selector.
+
+        Returns:
+            A list of matched gateway resources.
+        """
+        return list(
```

Comment:
Out of curiosity, I'm wondering if wrapping the value in list from `self._client.list` is needed :D (i'm guessint list() would return a list)

Reply:
Yes, I had to add `list()` for the linter to pass since apparently `client.list()` returns an iterator object rather than a list (which is indeed not intuitive at all :sweat_smile: )